src/lib.rs

```diff
@@ -177,13 +177,53 @@ impl Default for SweepStrategy {
 /// parallel execution, but for it to work, your query key/value types
 /// must also be `Send`, as must any additional data in your database.
 pub trait ParallelDatabase: Database + Send {
-    /// Creates a copy of this database destined for another
-    /// thread. See also `Runtime::fork`.
+    /// Creates a second handle to the database that holds the
+    /// database fixed at a particular revision. So long as this
+    /// "frozen" handle exists, any attempt to [`set`] an input will
+    /// block.
     ///
-    /// **Warning.** This second handle is intended to be used from a
+    /// [`set`]: struct.QueryTable.html#method.set
+    ///
+    /// This is the method you are meant to use most of the time in a
+    /// parallel setting where modifications may arise asynchronously
+    /// (e.g., a language server). In this context, it is common to
+    /// wish to "fork off" a copy of the database performing some
+    /// series of queries in parallel and arranging the results. Using
+    /// this method for that purpose ensures that those queries will
+    /// see a consistent view of the database (it is also advisable
+    /// for those queries to use the [`is_current_revision_canceled`]
+    /// method to check for cancellation).
+    ///
+    /// [`is_current_revision_canceled`]: struct.Runtime.html#method.is_current_revision_canceled
+    ///
+    /// # Deadlock warning
+    ///
+    /// This second handle is intended to be used from a separate
+    /// thread. Using two database handles from the **same thread**
+    /// can lead to deadlock.
+    fn fork(&self) -> Frozen<Self> {
+        Frozen::new(self)
+    }
+
+    /// Creates another handle to this database destined for another
+    /// thread. You are meant to implement this by cloning a second
+    /// handle to all all the state in the database; to clone the
+    /// salsa runtime, use the [`Runtime::fork_mut`] method.
+    ///
+    /// Note to users: you only need this method if you plan to be
+    /// setting inputs on the other thread. Otherwise, it is preferred
+    /// to use [`fork`], which will given back a frozen database fixed
+    /// at the current revision.
+    ///
+    /// # Deadlock warning
+    ///
+    /// This second handle is intended to be used from a
     /// separate thread. Using two database handles from the **same
     /// thread** can lead to deadlock.
-    fn fork(&self) -> Self;
+    ///
+    /// [`Runtime::fork_mut`]: https://docs.rs/salsa/0.7.0/salsa/struct.Runtime.html#method.fork_mut
+    /// [`fork`]: trait.ParallelDatabase.html#tymethod.fork
+    fn fork_mut(&self) -> Self;
```

Comment:
Do we have a use-case for setting from different threads?

I see that we *could* do it, but *should* we? Perhaps it makes sense to keep API minimal until someone comes along with a feature request?

Reply:
I don't have a particular use case but I don't see a way to make `fork` work without this method.

Reply:
In particular, I'm trying to model this API on the usual "rust mode", in which you have:

- `T` -- a type with full privileges
- `&T` -- here, `Frozen<T>`, a view on `T` with reduced privileges

And the point is that to implement `fork` which gives you the `Frozen<T>`, I  need a way to get another `T`

Reply:
Another question is: does supporting `fork_mut` cause a *problem* for some use case we want to support? I could *imagine* that it might at least make some things more awkward: it requires that the main `DB` type be "forkable" to other threads. So if you do have the model where there is a "main" thread with full access and "helper" threads with frozen access, maybe you'd like this 'main thread" to have some fields that they don't have.

This kind of corresponds to `String` (which has a capacity) and `str` (which does not), or the usual `PathBuf`/`Path` naming. So basically you could imagine a `DbBuf` type and a `Db` type. But I am wary because -- as those analogies may suggest -- that is another "Level up" in complexity for people to deal with.

I'd like to have a good use case for introducing a "buf-split". I am *mildly* skeptical that there is one -- you could always have a `Full<DB>` wrapper that adds the extra fields, after all. Moreover, you ought not to be adding shared, mutable state that *queries* can access since it would bypass the revision system and just be a footgun (and thus I'm not inclined to make it any easier).

So, put another way, does the `fork_mut` method do any **harm**?

Reply:
>And the point is that to implement fork which gives you the Frozen<T>, I need a way to get another T

That's a good point why we really need it. And no, I don't see any obvious problems with fork_mut.

Reply:
One thing I was wondering is whether there might be a better name than `fork_mut`. It feels like a qualitatively different operation than `fork`, in some sense, in that what you get out not only permits mutation but lacks the "single, consistent revision" property. (Of course, those are linked, but still.)

Reply:
Thinking more about it, I am wondering if it *is* worth trying to expose the "buf/view" split somehow. I think one of the main effects would be that the `salsa::Database` trait would represent *views* and we would add a separate `salsa::DatabaseBuf: salsa::Database` trait for the "backing buffer".. maybe? I have to think on it a bit. 

The main thing we would want, I think, is that:

- `Database` and `ParallelDatabase` support get queries and `fork`
- `DatabaseBuf` and `ParallelDatabaseBuf` add in the `query_mut(&mut self, query: Q)` and `fork_mut(&mut self)` operations

I think that the way this would work is that the runtime would also offer `fork(&self, from_db: &DB) -> Self` and `fork_mut(&mut self, from_db: &mut DB) -> Self` operations. 

Importantly, you would **not** have to implement `fork_mut` at all -- maybe this means we can even cut it out? But you would have to implement the `fork`, which -- since we want it to return a `Frozen<DB>` type I think -- is a *touch* awkward. You would do something like:

```rust
fn fork(&self) -> Frozen<Self> {
    Frozen::new(MyDatabaseType {
        runtime: self.runtime.fork(self),
        other_fields: self.other_fields.clone(),
    })
}
```

The advantages of this are:

- We do not need `fork_mut`, we can cut that for now altogether
- This implies we can **always** use `read_recursive` for the "query lock", because there is only one thread that can both get/set -- all other threads are using forked databases, which are already holding the read lock

The main disadvantage is that the user has to write the call to `Frozen::new` themselves.

Reply:
Er, that wound up in a slightly different place than I originally imagined, in that it doesn't have the buf/view split -- you still have `DB` and `Frozen<DB>` as the buf/view types. But it seems... good? Still not sure what a nice Buf/View split would look like of it that might feel slicker, seems like it must be strictly more work.

Reply:
(Maybe rename `Frozen<DB>` to `Forked<DB>`? Or `fork` to `freeze`? Or `fork_view` and `View`?)

Reply:
Hm, so I see various designs here, but neither strikes me as significantly better/worse than others. In Rust analyzer, I have a design where get/set parts are completely separate: you can't `get` from a "mutable" struct: I have an `AnalysisHost` with three methods, `new`, `apply_changes`, `snpahost`, and `Analysis` (returned by a `snapshot`), which has a bazillion of `get_foo` methods.  However, in this setup `apply_change` internally uses `.get` methods (change is a diff, so you need to know the current state to apply it). 


Unrelated, but maybe `Readonly<DB>` or `Snapshot<DB>`?